cpp/dolfinx/function/Function.cpp

```diff
@@ -101,49 +99,57 @@ Function Function::collapse() const
 
   // Create new vector
   assert(function_space_new);
-  la::PETScVector vector_new = create_vector(*function_space_new);
-
-  // Wrap PETSc vectors using Eigen
-  la::VecReadWrapper v_wrap(_vector.vec());
-  Eigen::Map<const Eigen::Matrix<PetscScalar, Eigen::Dynamic, 1>> x_old
-      = v_wrap.x;
-  la::VecWrapper v_new(vector_new.vec());
-  Eigen::Map<Eigen::Matrix<PetscScalar, Eigen::Dynamic, 1>> x_new = v_new.x;
+  auto vector_new = std::make_shared<la::Vector<PetscScalar>>(
+      function_space_new->dofmap()->index_map);
 
   // Copy values into new vector
+  const Eigen::Matrix<PetscScalar, Eigen::Dynamic, 1>& x_old = _x->array();
+  Eigen::Matrix<PetscScalar, Eigen::Dynamic, 1>& x_new = vector_new->array();
   for (std::size_t i = 0; i < collapsed_map.size(); ++i)
   {
     assert((int)i < x_new.size());
     assert(collapsed_map[i] < x_old.size());
     x_new[i] = x_old[collapsed_map[i]];
   }
 
-  return Function(function_space_new, vector_new.vec());
+  return Function(function_space_new, vector_new);
 }
 //-----------------------------------------------------------------------------
 std::shared_ptr<const FunctionSpace> Function::function_space() const
 {
   return _function_space;
 }
 //-----------------------------------------------------------------------------
-la::PETScVector& Function::vector()
+Vec Function::vector() const
 {
-  // Check that this is not a sub function.
+  // Check that this is not a sub function
   assert(_function_space->dofmap());
   assert(_function_space->dofmap()->index_map);
-  if (_vector.size()
+  if (_x->map()->block_size() * _x->map()->size_global()
       != _function_space->dofmap()->index_map->size_global()
              * _function_space->dofmap()->index_map->block_size())
   {
     throw std::runtime_error(
         "Cannot access a non-const vector from a subfunction");
   }
 
+  // Check that data type is the same as the PETSc build
+  // if constexpr (std::is_same<T, PetscScalar>::value)
+  // {
+  if (!_vector)
+  {
+    _vector = create_ghosted_vector(*_function_space->dofmap()->index_map,
+                                    _x->array());
+  }
   return _vector;
+  // }
```

Comment:
Leftover commented code.

Reply:
It's deliberate - will be activated immediately after OR.